presto-orc/src/main/java/com/facebook/presto/orc/OrcWriteValidation.java

```diff
@@ -236,35 +255,146 @@ public void validateRowGroupStatistics(OrcDataSourceId orcDataSourceId, long str
+                }
+            }
+        }
+
+        // return original stats excluding the flat map value stats for empty flat maps
+        return stats.entrySet().stream()
+                .filter(e -> !excludedNodes.contains(e.getKey()))
+                .collect(toImmutableMap(Entry::getKey, Entry::getValue));
+    }
+
+    private Map<Integer, ColumnStatistics> aggregateRowGroupStatisticsFromRowIndex(
+            OrcDataSourceId orcDataSourceId,
+            Map<StreamId, List<RowGroupIndex>> actualRowGroupStatistics,
+            long stripeOffset,
+            int rowGroupIndex)
+    {
+        // flattened nodes might have multiple ROW_INDEX with the same column, but different sequences
+        // aggregate such statistics before the validation
+        Map<Integer, List<ColumnStatistics>> actualColumnStatisticsByColumn = new HashMap<>();
+        for (Entry<StreamId, List<RowGroupIndex>> entry : actualRowGroupStatistics.entrySet()) {
+            int column = entry.getKey().getColumn();
+            ColumnStatistics actual = entry.getValue().get(rowGroupIndex).getColumnStatistics();
+            List<ColumnStatistics> aggregateStats = actualColumnStatisticsByColumn.computeIfAbsent(column, (key) -> new ArrayList<>());
+            aggregateStats.add(actual);
+
+            // Regular nodes have only 1 ColumnStatistics in the ROW_INDEX, flattened nodes
+            // might have zero or more column statistics.
+            if (aggregateStats.size() != 1 && !flattenedValueNodes.contains(column)) {
+                throw new OrcCorruptionException(
+                        orcDataSourceId,
+                        "Unexpected multiple column statistics for node %s in row group %s in stripe at offset %s",
+                        column,
+                        rowGroupIndex,
+                        stripeOffset);
+            }
+        }
+
+        return actualColumnStatisticsByColumn.entrySet().stream()
+                .collect(Collectors.toMap(Entry::getKey, entry -> mergeColumnStatistics(entry.getValue())));
+    }
+
+    private Map<Integer, Integer> getFlattenedKeyToMapNodes(Set<Integer> flattenedNodes, List<OrcType> orcTypes)
+    {
+        ImmutableMap.Builder<Integer, Integer> keyNodeToMapNode = ImmutableMap.builder();
+        flattenedNodes.forEach(mapNodeIndex -> keyNodeToMapNode.put(orcTypes.get(mapNodeIndex).getFieldTypeIndex(0), mapNodeIndex));
+        return keyNodeToMapNode.build();
+    }
+
+    private Map<Integer, Set<Integer>> getFlattenedMapToValueNodes(Set<Integer> flattenedNodes, List<OrcType> orcTypes)
+    {
+        ImmutableMap.Builder<Integer, Set<Integer>> keyNodeToMapNode = ImmutableMap.builder();
+        flattenedNodes.forEach(mapNodeIndex -> keyNodeToMapNode.put(mapNodeIndex, getFlattenedValueNodes(ImmutableSet.of(mapNodeIndex), orcTypes)));
+        return keyNodeToMapNode.build();
+    }
+
+    private Set<Integer> getFlattenedValueNodes(Set<Integer> flattenedNodes, List<OrcType> orcTypes)
```

Comment:
Should this method also take the mapNodeIndex ? Also the values are consecutive for a node, so you can put all the integers in a range.

Reply:
> Should this method also take the mapNodeIndex ?

Can you clarify the `take`? You mean return mapNodeIndex in the set? No need for this, we are fine with just the value nodes.

> Also the values are consecutive for a node, so you can put all the integers in a range.

See my answer about +1 for the key node. This is true only for OrcTypes created from Presto Types. To me it's safer to walk the tree instead of making assumptions.

Reply:
My bad the input to this method is ImmutableSet.of(mapNodeIndex), I assumed it takes in flattenedNodes again.